account-compression/sdk/src/events/index.ts

```diff
@@ -39,10 +46,79 @@ export function deserializeApplicationDataEvent(
     .toFixedFromData(data, 0)
     .read(data, 0);
   switch (event.__kind) {
-    case 'ApplicationData': {
+    case "ApplicationData": {
       return event.fields[0];
     }
     default:
-      throw Error('Unable to decode buffer as ApplicationDataEvent');
+      throw Error("Unable to decode buffer as ApplicationDataEvent");
   }
 }
+
+/**
+ * Helper function to extract the ChangeLogEvent V1 events from a TransactionResponse
+ * @param txResponse - TransactionResponse from the `@solana/web3.js`
+ * @param programId - PublicKey of the program (aka `programId`) that utilized the leaf on the tree
+ * @param noopProgramId - program id of the noop program used (default: `SPL_NOOP_PROGRAM_ID`)
+ * @returns
+ */
+export function getAllChangeLogEventV1FromTransaction(
+  txResponse: TransactionResponse | VersionedTransactionResponse,
+  programId: PublicKey,
+  noopProgramId: PublicKey = SPL_NOOP_PROGRAM_ID
+): ChangeLogEventV1[] {
+  // ensure a transaction response was provided
+  if (!txResponse) throw Error("No txResponse provided");
+
+  // flatten the array of all account keys (e.g. static, readonly, writable)
+  const accountKeys = txResponse.transaction.message
+    .getAccountKeys()
+    .keySegments()
+    .flat();
+
+  // find the correct index of the `programId` instruction
+  const relevantIndex =
+    txResponse.transaction.message.compiledInstructions.findIndex(
+      (instruction) =>
+        accountKeys[instruction.programIdIndex].toBase58() ===
+        programId.toBase58()
+    );
+
+  // locate the noop's inner instructions called via cpi from `programId`
+  const relevantInnerIxs = txResponse!.meta?.innerInstructions?.[
+    relevantIndex
+  ].instructions.filter((instruction) => {
+    return (
+      accountKeys[instruction.programIdIndex].toBase58() ===
+      noopProgramId.toBase58()
+    );
+  });
+
+  // when no valid noop instructions are found, throw an error
+  if (!relevantInnerIxs || relevantInnerIxs.length == 0)
+    throw Error("Unable to locate any noop cpi instructions");
+
+  let changeLogEvents: ChangeLogEventV1[] = [];
+
+  /**
+   * note: the ChangeLogEvent V1 is expected to be at position `1`,
+   * and normally expect only 2 `relevantInnerIx`
+   * so this sort method is more efficient for most uses cases
+   */
+  for (let i = relevantInnerIxs.length - 1; i >= 0; i--) {
+    try {
+      changeLogEvents.push(
+        deserializeChangeLogEventV1(
+          Buffer.from(bs58.decode(relevantInnerIxs[i]?.data!))
+        )
+      );
+    } catch (__) {
+      // this noop cpi is not a changelog event. do nothing with it.
+    }
+  }
+
+  // when no changeLogEvents were found, throw an error
+  if (changeLogEvents.length == 0)
+    throw Error("Unable to locate any `ChangeLogEventV1` events");
```

Comment:
I think this should just be an empty array, rather than an Error. Someone might use this to scan if they need to store transactions, and it might be annoying to catch an error for a normal output case.